feat(auth): Implement updated design for regional access boundary

nbayati opened the pull request:
Bring RAB (formerly known as trust boundaries) up to date with the design changes:
- Decouple RAB fetching from token refresh and use api calls as a trigger to fetch RAB.
- Make fetching async, non blocking, and fail open
- remove no-op signal
- Implement a 6 hour TTL with a soft expiry and proactive refresh

nbayati committed:
Remove no-op signal and logic.

nbayati committed:
Refactor to regional access boundary and use new env variable

nbayati committed:
feat: decouple RAB fetch from token refresh and make it async, nonblocking and fail open.

nbayati committed:
Refactor RAB unit tests

nbayati committed:
Update copyright date to 2026.

nbayati committed:
Implement soft expiry logic and simplify the env var feature gate.

nbayati committed:
Fixed blank lines that contained trailing tabs/spaces.

nbayati committed:
Expand cooldown to include errors on constructing the lookup URL.

nbayati committed:
trigger sync

nbayati committed:
feat: Make retry mechanism configurable and apply custom options to regional access boundary lookups.

nbayati committed:
fix formatting on retry test

quartzmo commented on auth/internal/regionalaccessboundary/regional_access_boundary.go:
```diff
@@ -0,0 +1,458 @@
+	p.cooldownExpiry = time.Now().Add(p.cooldownDuration)
+
+	// Exponential backoff for cooldown, up to cacheTTL max (6 hours)
+	nextCooldown := p.cooldownDuration * 2
```
The simple doubling (`cooldownDuration * 2`) here is missing a random jitter. This could cause cyclic thundering herds on expiry if many clients fail synchronously. 

`auth/internal/retry/retry.go` uses `math/rand` with `1 + rand.Int63n(int64(b.cur))` for its jitter mechanism. We could adopt a similar approach for the cooldown expiry.

quartzmo commented on auth/credentials/detect.go:
```diff
@@ -142,7 +142,7 @@ func DetectDefault(opts *DetectOptions) (*auth.Credentials, error) {
 		return nil, err
 	}
-	trustBoundaryEnabled, err := trustboundary.IsEnabled()
+	trustBoundaryEnabled, err := regionalaccessboundary.IsEnabled()
```
The local var `trustBoundaryEnabled` should also be renamed.

nbayati committed:
Pass service endpoint to SetAuthMetadata for regional checks and skipping RAB

nbayati committed:
fix: Make URL parsing more robust when checking for regional endpoints

nbayati committed:
Add random jitter to cooldownExpiry for RAB lookup

nbayati committed:
Use strict anchors for external account audience regexes

quartzmo reviewed (changes requested):
```
chrisdsmith@chrisdsmith:~/projects/rab-review/repos/google-cloud-go/auth$ go test --race -run TestDataProvider_GetHeaderValue ./internal/regionalaccessboundary/...
==================
WARNING: DATA RACE
Write at 0x00c0000145a0 by goroutine 15:
  cloud.google.com/go/auth/internal/regionalaccessboundary.(*mockConfigProvider).GetRegionalAccessBoundaryEndpoint()
      /usr/local/google/home/chrisdsmith/projects/rab-review/repos/google-cloud-go/auth/internal/regionalaccessboundary/regional_access_boundary_test.go:593 +0x44
  cloud.google.com/go/auth/internal/regionalaccessboundary.(*DataProvider).fetchAsync()
      /usr/local/google/home/chrisdsmith/projects/rab-review/repos/google-cloud-go/auth/internal/regionalaccessboundary/regional_access_boundary.go:320 +0xef
  cloud.google.com/go/auth/internal/regionalaccessboundary.(*DataProvider).GetHeaderValue.gowrap3()
      /usr/local/google/home/chrisdsmith/projects/rab-review/repos/google-cloud-go/auth/internal/regionalaccessboundary/regional_access_boundary.go:307 +0x5d

Previous read at 0x00c0000145a0 by goroutine 14:
  cloud.google.com/go/auth/internal/regionalaccessboundary.TestDataProvider_GetHeaderValue.func1.1()
      /usr/local/google/home/chrisdsmith/projects/rab-review/repos/google-cloud-go/auth/internal/regionalaccessboundary/regional_access_boundary_test.go:667 +0x371
  testing.tRunner()
      /usr/lib/go-1.24/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /usr/lib/go-1.24/src/testing/testing.go:1851 +0x44

Goroutine 15 (running) created at:
  cloud.google.com/go/auth/internal/regionalaccessboundary.(*DataProvider).GetHeaderValue()
      /usr/local/google/home/chrisdsmith/projects/rab-review/repos/google-cloud-go/auth/internal/regionalaccessboundary/regional_access_boundary.go:307 +0x8ee
  cloud.google.com/go/auth/internal/regionalaccessboundary.TestDataProvider_GetHeaderValue.func1.1()
      /usr/local/google/home/chrisdsmith/projects/rab-review/repos/google-cloud-go/auth/internal/regionalaccessboundary/regional_access_boundary_test.go:661 +0x248
  testing.tRunner()
      /usr/lib/go-1.24/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /usr/lib/go-1.24/src/testing/testing.go:1851 +0x44

Goroutine 14 (running) created at:
  testing.(*T).Run()
      /usr/lib/go-1.24/src/testing/testing.go:1851 +0x8f2
  cloud.google.com/go/auth/internal/regionalaccessboundary.TestDataProvider_GetHeaderValue.func1()
      /usr/local/google/home/chrisdsmith/projects/rab-review/repos/google-cloud-go/auth/internal/regionalaccessboundary/regional_access_boundary_test.go:656 +0x244
  testing.tRunner()
      /usr/lib/go-1.24/src/testing/testing.go:1792 +0x225
  testing.(*T).Run.gowrap1()
      /usr/lib/go-1.24/src/testing/testing.go:1851 +0x44
==================
--- FAIL: TestDataProvider_GetHeaderValue (0.19s)
    --- FAIL: TestDataProvider_GetHeaderValue/Regional_endpoint_skip (0.16s)
        --- FAIL: TestDataProvider_GetHeaderValue/Regional_endpoint_skip/Injected_Query_Param (0.01s)
            testing.go:1490: race detected during execution of test
FAIL
FAIL	cloud.google.com/go/auth/internal/regionalaccessboundary	0.210s
FAIL
```

The good news is that this is a test-only race, not a production bug.

The race detector is complaining about the mockConfigProvider. The background goroutine increments a counter with m.endpointCallCount++, while the test thread is simultaneously reading it in a polling loop to check if the fetch was triggered.

I think you need to protect the call counts in the mock with a sync.Mutex and Lock/Unlock calls in the mockConfigProvider methods.

nbayati committed:
test: add mutex to mockConfigProvider to prevent races in tests

quartzmo reviewed (approved):
Great job!!!

nbayati commented on auth/internal/regionalaccessboundary/regional_access_boundary.go:
```diff
@@ -0,0 +1,473 @@
+		if h, _, err := net.SplitHostPort(host); err == nil {
+			host = h
+		}
+		if host == "rep.googleapis.com" || strings.HasSuffix(host, ".rep.googleapis.com") ||
```
I decided not to add checks against iam and sts endpoints here, mainly because:
1. the calls to iam and sts don't necessarily go through this flow. For example [here](https://github.com/googleapis/google-cloud-go/blob/b7c7bf874b7c12c128ca31bff5c00cf7d837267e/auth/credentials/internal/externalaccount/externalaccount.go#L317), the STS token exchange logic uses creates its own set of headers and uses the raw Client: tp.client to execute the request, which means it won't go through the GetHeaderValue flow.  
2. the backend will ignore the header if it's presented for an iam or sts request. Our design says we don't have to go out of our way to ensure the header is attached to all the iam and sts calls, but if it is attached, it's not an issue. 
3. it's different from the regional endpoint check, since we want to optimize by not fetching the header for regional endpoints, but if a credential is calling iam or sts, they'll be calling other endpoints next that might be regional or global. Adding a check here won't save performance, and it might make the logic brittle and hard to maintain if we have to extract the target and check against iam and sts URLs.

lsirac commented on auth/internal/regionalaccessboundary/regional_access_boundary.go:
```diff
@@ -0,0 +1,473 @@
+		token.Metadata = make(map[string]interface{})
+	}
+
+	token.Metadata[ProviderKey] = p
```
I think we should avoid mutating the token returned by the wrapped provider here. The TokenProvider contract treats returned tokens as read-only, and some providers return cached/shared token pointers.

The GCE path looks especially risky because computeTokenProvider returns a cached provider, and compute tokens use the package-level computeTokenMetadata map. So adding ProviderKey in place can mutate shared metadata, race between concurrent callers, or leak this RAB provider pointer across otherwise separate compute credential instances.

Could we return a shallow copy of the token and a cloned Metadata map, then add ProviderKey to the cloned metadata before returning?

nbayati replied:
Great catch, thanks for flagging this!

lsirac commented on auth/internal/regionalaccessboundary/regional_access_boundary.go:
```diff
@@ -0,0 +1,473 @@
+// GetRegionalAccessBoundaryEndpoint constructs the Regional Access Boundary lookup URL for a GCE environment.
+// It uses cached service account email after the first call.
+func (g *GCEConfigProvider) GetRegionalAccessBoundaryEndpoint(ctx context.Context) (string, error) {
+	g.saOnce.Do(func() { g.fetchSA(ctx) })
```
fetchAsync now applies cooldown when endpoint construction fails, but here a first-call EmailWithContext failure sets saEmailErr permanently. Later cooldown retries will keep returning the same cached error.

Is this intentional? Could we avoid caching failed service-account-email lookups, or only mark the email initialized after a successful lookup? Otherwise this path defeats the retry/cooldown behavior the PR adds.

nbayati replied:
Good catch, I think the idea of only trying once was from when the call was blocking. I've updated the code to make it cache the successful result, but not the error. This way it can try to get the SA again after the cooldown.

lsirac commented on auth/grpctransport/grpctransport.go:
```diff
@@ -326,6 +326,7 @@ func dial(ctx context.Context, secure bool, opts *Options) (*grpc.ClientConn, er
 				creds:                creds,
 				metadata:             metadata,
 				clientUniverseDomain: opts.UniverseDomain,
+				endpoint:             transportCreds.Endpoint,
```
I think DirectPath will still lose the endpoint RAB needs for the regional endpoint skip. 

After this, `configureDirectPath` can replace the existing options with new `grpcCredentialsProvider` instances in `directpath.go`. Those providers currently only set `creds`, so `c.endpoint` is empty when `GetRequestMetadata` calls `headers.SetAuthMetadata(token, c.endpoint, metadata)`. That means `GetHeaderValue` can't recognize `*.rep.googleapis.com` and skip `x-allowed-locations` for regional endpoints.

Can the DirectPath providers also carry the original service endpoint, ideally before adding resolver prefixes?

nbayati replied:
Done!

lsirac commented on auth/internal/transport/headers/headers.go:
```diff
@@ -31,31 +37,26 @@ func SetAuthHeader(token *auth.Token, req *http.Request) {
-		if tbd, ok := data.(internal.TrustBoundaryData); ok {
-			return tbd.TrustBoundaryHeader()
+	if provider, ok := token.Metadata[regionalaccessboundary.ProviderKey].(regionalAccessBoundaryProvider); ok {
+		if headerVal := provider.GetHeaderValue(context.Background(), reqURL, token); headerVal != "" {
```
I think this should preserve the caller’s context instead of using context.Background().

In the gRPC path, GetRequestMetadata receives the per-RPC context, but SetAuthMetadata drops it here. GetHeaderValue synchronously calls GetUniverseDomain before it can return, and on GCE that can hit metadata under sync.Once.

So a per-RPC auth metadata call can ignore the RPC deadline or cancellation and block even though the RAB fetch path is intended to be async/fail-open.

Could SetAuthMetadata take a context and pass the gRPC request context through to GetHeaderValue?

nbayati replied:
Good call, done!

lsirac commented on auth/internal/regionalaccessboundary/regional_access_boundary.go:
```diff
@@ -0,0 +1,473 @@
+
+	// Skip lookup for non-default universe domains.
+	uniDomain, err := p.configProvider.GetUniverseDomain(ctx)
+	if err == nil && uniDomain != "" && uniDomain != internal.DefaultUniverseDomain {
```
Should universe-domain lookup errors fail open by skipping RAB?

nbayati replied:
Yes, fixed!

lsirac commented on auth/internal/regionalaccessboundary/external_accounts_config_providers.go:
```diff
@@ -21,13 +21,13 @@ import (
 var (
-	workforceAudiencePattern = regexp.MustCompile(`//iam\.([^/]+)/locations/global/workforcePools/([^/]+)`)
-	workloadAudiencePattern  = regexp.MustCompile(`//iam\.([^/]+)/projects/([^/]+)/locations/global/workloadIdentityPools/([^/]+)`)
+	workforceAudiencePattern = regexp.MustCompile(`^//iam\.([^/]+)/locations/global/workforcePools/([^/]+)$`)
```
This needs to include the provider portion. Otherwise it only matches `//iam.googleapis.com/projects/123/locations/global/workloadIdentityPools/my-pool`
but not `//iam.googleapis.com/projects/123/locations/global/workloadIdentityPools/my-pool/providers/my-provider`, which won't work.

Should be anchoring the full expected shape: ^//iam\.([^/]+)/projects/([^/]+)/locations/([^/]+)/workloadIdentityPools/([^/]+)/providers/([^/]+)$

Also, could we add/reuse a shared external-account audience parser instead of having a RAB-specific regex here?

nbayati replied:
Ah thanks for flagging this. A recent commit had broken this flow and I hadn't realized it. 
I decided against creating a shared parser for now, because it'd require creating a new package under auth that both RAB and other external account logics could use. I'm going to have a follow up PR to enable RAB, and I'll see if I can cleanly refactor it then.

lsirac commented on auth/internal/regionalaccessboundary/external_accounts_config_providers_test.go:
These RAB parser tests are using shortened pool resource names, but external-account credential `audience` values include the provider segment, e.g. `.../workloadIdentityPools/my-pool/providers/my-provider` and `.../workforcePools/my-pool/providers/my-provider`.

We need to update this to use the full provider-resource audience strings. Otherwise the tests incorrectly pass for a shape that `filetypes.go` will not normally pass into `NewExternalAccountConfigProvider`.

nbayati replied:
Done!

lsirac commented on auth/internal/regionalaccessboundary/regional_access_boundary_test.go:
```diff
@@ -649,290 +628,151 @@ func (m *mockTokenProvider) Token(ctx context.Context) (*auth.Token, error) {
 
 func TestDataProvider_Token(t *testing.T) {
 	ctx := context.Background()
+	baseToken := &auth.Token{Value: "base-token"}
```
Could we extend this test to assert that Token() does not mutate the token returned by the wrapped provider?

Right now this only checks that the returned token has ProviderKey, so the test would still pass if baseToken.Metadata is mutated in place.

Let's pre-populate baseToken.Metadata with a sentinel value, call provider.Token(ctx), then assert:
- the original baseToken.Metadata does not contain ProviderKey
- the returned token preserves the existing metadata entries
- the returned metadata map is not the same map as baseToken.Metadata

nbayati replied:
Done!

lsirac commented on auth/internal/regionalaccessboundary/regional_access_boundary_test.go:
```diff
@@ -649,290 +628,151 @@ func (m *mockTokenProvider) Token(ctx context.Context) (*auth.Token, error) {
+		}
+	})
+
+	t.Run("Endpoint_fetch_failure_sets_cooldown", func(t *testing.T) {
```
Let's add a GCE-specific transient-failure regression test here

lsirac commented on auth/internal/transport/headers/headers_test.go:
```diff
@@ -82,22 +79,22 @@ func TestSetAuth(t *testing.T) {
 			// Test gRPC
 			gotMeta := make(map[string]string)
-			SetAuthMetadata(token, gotMeta)
+			SetAuthMetadata(token, "https://example.com/v1", gotMeta)
```
Could we add coverage that the caller context reaches the RAB provider in the gRPC metadata path?

The current mock provider only returns a static value, so this test would still pass if SetAuthMetadata uses context.Background().

After changing SetAuthMetadata to accept a context, this test could pass a context with a sentinel value or cancellation/deadline and assert the mock provider saw that same context. Ideally there would also be a small grpcCredentialsProvider.GetRequestMetadata test proving the per-RPC context is passed through.

nbayati replied:
I have addressed your suggestions regarding context propagation in the gRPC path:

I updated SetAuthMetadata in `headers.go` to accept a context.Context and passed it through to GetHeaderValue instead of using context.Background().

I extended TestSetAuth in `headers_test.go` by passing a context with a custom key/value to SetAuthMetadata and asserting that the mock provider receives that same context.

I also added a new test, TestGrpcCredentialsProvider_GetRequestMetadata_Context, in `grpctransport_test.go`. This test creates a mock state with a custom context and verifies that GetRequestMetadata passes that per-RPC context all the way down to the RAB provider's GetHeaderValue method.

lsirac commented on auth/internal/retry/retry.go:
```diff
@@ -22,10 +22,6 @@ import (
-)
-
 var (
 	syscallRetryable = func(error) bool { return false }
```
This is hardcoded to always return false, so the check on line 149 in this file and transient socket errors from the iamcredentials client fall through as non-retryable on every platform.

The compute/metadata package has the same stub in compute/metadata/retry.go but pairs it with compute/metadata/retry_linux.go, which swaps in a real implementation under a build tag. 

Could we add a matching auth/internal/retry/retry_linux.go, or give the default implementation in auth/internal/retry/retry.go something real to check so the branch on line 149 is actually reachable?

nbayati replied:
Great catch, thank you!

lsirac commented on auth/credentials/filetypes.go:
```diff
@@ -306,19 +306,19 @@ func handleImpersonatedServiceAccount(f *credsfile.ImpersonatedServiceAccountFil
-	targetSAConfig := trustboundary.NewServiceAccountConfigProvider(targetSAEmail, ud)
-	return trustboundary.NewProvider(opts.client(), targetSAConfig, opts.logger(), tp)
+	targetSAConfig := regionalaccessboundary.NewServiceAccountConfigProvider(targetSAEmail, ud)
+	return regionalaccessboundary.NewProvider(opts.client(), targetSAConfig, opts.logger(), tp)
```
I think this target-credential RAB wrapping can interact badly with the source credential built earlier in this function.

Above, sourceTP is built through fileCredentials(f.CredSource, &sourceOpts). If RAB is enabled, that source credential may already be wrapped with its own RAB provider. Then the impersonation provider uses the source token for the internal generateAccessToken call, and SetAuthHeader can see the source token’s ProviderKey, start a source allowedLocations lookup, and potentially attach the source x-allowed-locations header to the IAMCredentials impersonation call.

But the boundary that should matter for user resource calls is the final impersonated service account, which is what this outer wrapping handles. The source-boundary lookup/header seems like extra lookup traffic and log noise at best, and it relies on IAMCredentials ignoring an unrelated source-boundary header.

Could we suppress RAB when building the impersonation source credential, or explicitly skip RAB header injection for IAMCredentials / STS internal auth plumbing calls?

nbayati replied:
Added a iam/sts check to the `GetHeaderValue` method to skip RAB lookup.

nbayati committed:
auth: address review feedback for Regional Access Boundary

- Fix GCE service account email caching to avoid caching failures permanently.
- Propagate the original service endpoint in DirectPath providers.
- Preserve caller context in the gRPC metadata path.
- Exclude IAM and STS hosts from RAB lookups to skip unnecessary calls and avoid log noise.
- Update external account audience regexes to support full provider paths.
- Add regression tests for transient GCE failures and context propagation.

lsirac commented on auth/internal/regionalaccessboundary/regional_access_boundary.go:
```diff
@@ -0,0 +1,500 @@
+
+// GetUniverseDomain retrieves the universe domain from the GCE metadata server.
+// It uses a cached value after the first call.
+func (g *GCEConfigProvider) GetUniverseDomain(ctx context.Context) (string, error) {
```
I think this will permanently cache transient universe-domain lookup failures via `udOnce` + `udErr`.

That means if the first metadata universe-domain lookup fails transiently, every later `GetHeaderValue` call will keep failing open forever for this provider, even after metadata recovers. This is the same as the earlier `saOnce`/`saEmailErr` issue that was fixed for service-account email lookup.

nbayati replied:
You're right. I think we have two options, either change the underlying method that fetches the UD not to permanently cache failures, or write our own code to fetch it. I need to see why it was written that way in the internal method, and will address it in my follow up PR.

lsirac commented on auth/internal/regionalaccessboundary/regional_access_boundary.go:
```diff
@@ -0,0 +1,500 @@
+	if !strings.Contains(reqURL, "://") {
+		reqURL = "https://" + reqURL
+	}
+	if u, err := url.Parse(reqURL); err == nil {
```
Could we add a small regression test for resolver-scheme endpoints, since url.Parse leaves Host empty there?

e.g. see endpoints in https://pkg.go.dev/google.golang.org/grpc#NewClient

nbayati replied:
Done!

lsirac reviewed (approved):
Great work!

nbayati committed:
fix: handle dns resolver scheme hostnames in regional access boundary and add regression tests

nbayati committed:
fix lint error
